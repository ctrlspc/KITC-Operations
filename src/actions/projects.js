
export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id, 
  updates
});
