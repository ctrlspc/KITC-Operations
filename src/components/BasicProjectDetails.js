import React from 'react';

export default ({title, description, projectManager, projectType}) => (
  <div>
    <h3>{title}</h3>
    <p>{description}</p>
    <p><span>Project Manager:</span> {projectManager.name}</p>
  </div>
);