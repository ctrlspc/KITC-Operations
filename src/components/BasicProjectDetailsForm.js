import React from 'react';
import _ from 'lodash';
export default class BasicProjectDetailsForm extends React.Component {
 
  constructor(props) {
    super(props);
    const project = this.props.project;
    this.state = {
      title: project ? project.title : '',
      description: project ? project.description : '',
      projectManager: project ? project.projectManager.uid  : '' ,
      projectType: project ? project.projectType : 'ext',
      error:''
    };
  };

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.projectManagers !== this.props.projectManagers) {
      if (this.state.projectManager === '' && !_.isEmpty(this.props.projectManagers[0])) {
        this.setState(() => ({projectManager:this.props.projectManagers[0].uid}));
      }
    }
  };

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({title}));
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onManagerChange = (e) => {
    const projectManager = e.target.value;
    this.setState(() => ({projectManager}));
  };

  onTypeChange = (e) => {
    const projectType = e.target.value;
    this.setState(() => ({projectType}));
  };

  onSubmit = (e) => {
    e.preventDefault();

    const projectManager = this.props.projectManagers.find((projectManager) => {
      return projectManager.uid === this.state.projectManager
    });
    if(!this.state.title || !this.state.description || !this.state.projectManager) {
      this.setState(() => ({error:'Please provide a title and project description'}));
    } else {

      this.setState(() => ({error:''}));
      this.props.onSubmit({
        title: this.state.title,
        description: this.state.description,
        projectType: this.state.projectType,
        projectManager
      });
    }
  };

  render () {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p> }
        <input 
          autoFocus
          className="text-input"
          type="text" 
          placeholder="Project Name"
          value={this.state.title}
          onChange={this.onTitleChange}
        />
        <textarea 
          className="textarea"
          placeholder="Project Description"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <div className="select-group">
          <div className="select-group__item">
            <h3 className="select-group__title">Project Manager</h3>
            <select 
              className="select-group__select"
              id="project-manager"
              onChange={this.onManagerChange}
              value={this.state.projectManager}
            >
              {this.props.projectManagers && this.props.projectManagers.map((item) => {
                return <option key={item.uid} value={item.uid}>{item.displayName}</option> 
              })}
            </select>
          </div>
          <div className="select-group__item">
            <h3 className="select-group__title">Project Type</h3>
            <select 
              className="select-group__select"
              id="project-type" 
              onChange={this.onTypeChange}
              value={this.state.projectType}
            >
              <option value="int">Internal</option>
              <option value="ext">External</option>
            </select>
          </div>
        </div>
        <div>
          <button className="button">Save Project</button>
        </div>
        
      </form>
    );
  };
};
