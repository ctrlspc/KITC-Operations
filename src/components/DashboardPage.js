import React from 'react';
import ProjectList from './ProjectList';
import { Link } from 'react-router-dom';

const DashboardPage = () => (

  <div>
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Projects Dashboard
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Create a Project</Link>
        </div>
      </div>
    </div>
    <div className="content-container">
      <ProjectList />
    </div>
  </div>
  
);

export default DashboardPage;