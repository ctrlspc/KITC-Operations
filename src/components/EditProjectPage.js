import React from 'react';
import { connect } from 'react-redux';
import ProjectForm from '../components/ProjectForm';
import { updateProject } from '../actions/projects';

export class EditProjectPage extends React.Component {

  onSubmit = (project) => {
    this.props.updateProject(this.props.project.id, project);
    this.props.history.push('/');
  }

  render () {
    return (
      <div className="content-container">
        <h1>Edit an existing Project</h1>
        <ProjectForm 
          onSubmit={this.onSubmit} 
          projectManagers={[
            {id:'1',name:'Jason Marshall'}, {id:'2', name:'Bec Lloyds'}
          ]}
          project={this.props.project}
        />
      </div>
    );
  };
};

const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
  updateProject: (id, project) => dispatch(updateProject(id, project))
})


export default connect(mapStateToProps,mapDispatchToProps)(EditProjectPage);