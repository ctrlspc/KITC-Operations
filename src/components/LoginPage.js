import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/iam';

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">KITC</h1>
      <p>Project Operations Tracker</p>
      <button className="button" onClick={startLogin}>Login with Google</button>
    </div>
    
  </div>
);

export const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);