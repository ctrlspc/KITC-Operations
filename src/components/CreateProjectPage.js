import React from 'react';
import { connect } from 'react-redux';

import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startAddProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class CreateProjectPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
  }

  onSubmit = (project) => {
    this.props.addProject(project);
    this.props.history.push('/');
  };

  render () {
    return (
      <div className="content-container">
        <h1>Create a new Project</h1>
        <BasicProjectDetailsForm 
          onSubmit={this.onSubmit} 
          projectManagers={this.props.users}
        />
      </div>
    );
  };
};

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  addProject: (project) => dispatch(startAddProject(project))
});

export const mapStateToProps = (state) => ({
  users: getActiveUsers(state)
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateProjectPage);


