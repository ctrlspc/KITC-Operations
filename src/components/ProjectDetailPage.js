import React from 'react';
import { connect } from 'react-redux';
import BasicProjectDetails from './BasicProjectDetails';
import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startUpdateProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class ProjectDetailPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
    this.state = {
      editBasicProjectDetails: !!props.editBasicProjectDetails
    };
  }

  onEditBasicProjectDetails = () => {
    this.setState(() => ({editBasicProjectDetails: true}));
  };

  onFinishedEditingBasicProjectDetails = (updates) => {
    this.props.updateProject(this.props.project.id, updates).then(() => {
      this.setState(() => ({editBasicProjectDetails: false}));
    });
  }

  render (){
    return (
      <div className="content-container">
        <div className="box-section">
          <div className="box-section__header">
            <h1>Basic Project Details</h1>
            {!this.state.editBasicProjectDetails && 
              <button className="button" onClick={this.onEditBasicProjectDetails}>Edit</button>
            }
          </div>
          <div className="box-section__body">
            {
              this.state.editBasicProjectDetails ? 
              (
                <BasicProjectDetailsForm 
                  onSubmit={this.onFinishedEditingBasicProjectDetails} 
                  projectManagers={this.props.users}
                  project={this.props.project}
                />
              ) : (
                <BasicProjectDetails {...this.props.project} onEditClick={this.onEditBasicProjectDetails}/>
              )
            }
          </div>
        </div>
      </div>
    );
  };
};

export const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id),
  users: getActiveUsers(state)
});

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  updateProject: (id, project) => dispatch(startUpdateProject(id, project))
})

export default connect(mapStateToProps, mapDispatchToProps)(ProjectDetailPage);