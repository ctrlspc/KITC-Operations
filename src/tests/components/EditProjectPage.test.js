import React from 'react';
import { shallow } from 'enzyme';
import { EditProjectPage } from '../../components/EditProjectPage';

import projects from '../fixtures/projects';

let updateProject, history, wrapper, project ;

beforeEach(() => {
  updateProject = jest.fn();
  project = projects[0];
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditProjectPage 
      updateProject={updateProject} 
      history={history}
      project={project}
    />);
});

test('should render the EditProjectPage component correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('ProjectForm').prop('onSubmit')(projects[0]);
  expect(updateProject).toHaveBeenLastCalledWith(projects[0].id, projects[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
});