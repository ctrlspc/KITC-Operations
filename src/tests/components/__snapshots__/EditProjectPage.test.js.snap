// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the EditProjectPage component correctly 1`] = `
<div
  className="content-container"
>
  <h1>
    Edit an existing Project
  </h1>
  <ProjectForm
    onSubmit={[Function]}
    project={
      Object {
        "description": "testing some great stuff",
        "id": "1",
        "projectManager": Object {
          "id": "1",
          "name": "Mr Test",
        },
        "projectType": "int",
        "title": "first test project",
      }
    }
    projectManagers={
      Array [
        Object {
          "id": "1",
          "name": "Jason Marshall",
        },
        Object {
          "id": "2",
          "name": "Bec Lloyds",
        },
      ]
    }
  />
</div>
`;
